include/aie/detail/emulation.hpp

```diff
@@ -11,16 +11,123 @@
 
 #include "common.hpp"
 #include "sync.hpp"
+#include "fiber_pool.hpp"
 
 namespace aie::detail {
 
-using device_tile_impl = device_tile_impl_fallback;
 using host_lock_impl = lock_impl_fallback;
 using device_lock_impl = lock_impl_fallback;
-using device_impl = device_impl_fallback;
-using host_tile_impl = host_tile_impl_fallback;
-using device_mem_handle_impl = device_mem_handle_impl_fallback;
 
+struct device_mem_handle_impl {
+  void memcpy_h2d(generic_ptr<void> dst, void* src, uint32_t size) {
+    std::memcpy(dst.ptr, src, size);
+  }
+  void memcpy_d2h(void* dst, generic_ptr<void> src, uint32_t size) {
+    std::memcpy(dst, src.ptr, size);
+  }
+};
+
+using device_mem_handle = device_mem_handle_adaptor<device_mem_handle_impl>;
+
+struct host_tile_impl;
+
+struct device_impl {
+  /// type-erased addresses of every tile's storage on the host.
+  struct tile_info {
+    void* mem;
+    host_tile_impl* tile;
+  };
+  std::vector<tile_info> tiles;
+  int sizeX;
+  int sizeY;
```

Comment:
We use `std::` style coding standard in this project
```suggestion
  int size_x;
  int size_y;
```